spring-batch-integration/src/main/java/org/springframework/batch/integration/partition/RemotePartitioningWorkerStepBuilder.java

```diff
@@ -0,0 +1,258 @@
+import org.springframework.batch.core.explore.JobExplorer;
+import org.springframework.batch.core.job.flow.Flow;
+import org.springframework.batch.core.partition.support.Partitioner;
+import org.springframework.batch.core.repository.JobRepository;
+import org.springframework.batch.core.step.StepLocator;
+import org.springframework.batch.core.step.builder.FlowStepBuilder;
+import org.springframework.batch.core.step.builder.JobStepBuilder;
+import org.springframework.batch.core.step.builder.PartitionStepBuilder;
+import org.springframework.batch.core.step.builder.SimpleStepBuilder;
+import org.springframework.batch.core.step.builder.StepBuilder;
+import org.springframework.batch.core.step.builder.TaskletStepBuilder;
+import org.springframework.batch.core.step.tasklet.Tasklet;
+import org.springframework.batch.repeat.CompletionPolicy;
+import org.springframework.beans.factory.BeanFactory;
+import org.springframework.integration.channel.NullChannel;
+import org.springframework.integration.dsl.IntegrationFlow;
+import org.springframework.integration.dsl.IntegrationFlows;
+import org.springframework.integration.dsl.StandardIntegrationFlow;
+import org.springframework.integration.dsl.context.IntegrationFlowContext;
+import org.springframework.messaging.MessageChannel;
+import org.springframework.transaction.PlatformTransactionManager;
+import org.springframework.util.Assert;
+
+/**
+ * Builder for a worker step in a remote partitioning setup. This builder
+ * creates an {@link IntegrationFlow} that:
+ *
+ * <ul>
+ *     <li>listens to {@link StepExecutionRequest}s coming from the master
+ *     on the input channel</li>
+ *     <li>invokes the {@link StepExecutionRequestHandler} to execute the worker
+ *     step for each incoming request. The worker step is located using the provided
+ *     {@link StepLocator}. If no {@link StepLocator} is provided, a {@link BeanFactoryStepLocator}
+ *     configured with the current {@link BeanFactory} will be used
+ *     <li>replies to the master on the output channel (when the master step is
+ *     configured to aggregate replies from workers). If no output channel
+ *     is provided, a {@link NullChannel} will be used (assuming the master side
+ *     is configured to poll the job repository for workers status)</li>
+ * </ul>
+ *
+ * @since 4.1
+ * @author Mahmoud Ben Hassine
+ */
+public class RemotePartitioningWorkerStepBuilder extends StepBuilder {
+
+	private static final String SERVICE_ACTIVATOR_METHOD_NAME = "handle";
+	private static final Log logger = LogFactory.getLog(RemotePartitioningWorkerStepBuilder.class);
+
+	private MessageChannel inputChannel;
+	private MessageChannel outputChannel;
+	private JobExplorer jobExplorer;
+	private StepLocator stepLocator;
+	private BeanFactory beanFactory;
+
+	/**
+	 * Initialize a step builder for a step with the given name.
+	 * @param name the name of the step
+	 */
+	public RemotePartitioningWorkerStepBuilder(String name) {
```

Comment:
Are we really allowed to create instances of this class directly, but not only via an appropriate `factory`?

Reply:
Yes you are allowed to use the builder directly, but you have to set the properties (that the factory sets) manually.

Reply:
OK. So, if we can instantiate builders directly without factory, then my previous comment about removing `public` modifier is wrong and end-user should be able to provide all the necessary options for the builder.